### PR TITLE
Fix Activities view

### LINKED ISSUE #1752: New Activity view + Comment
1. This activity indicator seems be too much 

Simulator Screen Shot - iPhone 13 Pro - 2021-11-11 at 19 06 28

2. How many time is called the API during the refresh .. ?

Schermata 2021-11-11 alle 19 16 36

3. The "Cancel" seem be useless

Simulator Screen Shot - iPhone 13 Pro - 2021-11-11 at 19 05 10

### PR DESCRIPTION
**Pull request by thisIsTheFoxe**
- Fix #1752 

TODO: 
- [x] DCO

### CODE REVIEW

File: iOSClient/Activity/NCActivity.swift [hunk 1]
```diff
@@ -351,69 +350,86 @@ extension NCActivity: UITableViewDataSource {
...
+            if let disptachGroup = disptachGroup {
+                disptachGroup.leave()
+            } else {
+                self.loadDataSource()
+            }
         }
-        
+    }
+    
+    func loadActivity(idActivity: Int, disptachGroup: DispatchGroup? = nil) {
```
Comment (marinofaggiana) [inline iOSClient/Activity/NCActivity.swift:430]: loadActivity is called always with disptachGroup(), so this arguments can't be optional (?)

File: iOSClient/Activity/NCActivity.swift [hunk 2]
```diff
@@ -351,69 +350,86 @@ extension NCActivity: UITableViewDataSource {
...
+        disptachGroup?.enter()
```
Comment (marinofaggiana) [inline iOSClient/Activity/NCActivity.swift:431]: remove ?

File: iOSClient/Activity/NCActivity.swift [hunk 3]
```diff
@@ -425,15 +441,11 @@ extension NCActivity {
                     NCManageDatabase.shared.addActivity(activities, account: account)
                 }
                 
-                NCUtility.shared.stopActivityIndicator()
-                
-                if errorCode == NCGlobal.shared.errorNotModified {
-                    self.canFetchActivity = false
+                if let disptachGroup = disptachGroup {
```
Comment (marinofaggiana) [inline iOSClient/Activity/NCActivity.swift:444]: replace with disptachGroup.leave()

File: iOSClient/Activity/NCActivity.swift (hunk 1 above)
Comment (thisIsTheFoxe) [reply iOSClient/Activity/NCActivity.swift:430]: Correct 👍

**Comment by marinofaggiana** [comment]
The load of old activity don't works, this capture view the last activity load after don't see anything but the activity via web site are much older:

Simulator Screen Shot - iPhone 13 - 2021-11-16 at 09 22 25

web:

Schermata 2021-11-16 alle 09 26 08

**Comment by marinofaggiana** [comment]
Note optimization: 

1. when start view immediately the item stored. (#now I must wait the loading ... )

2. when load the view reload always the first 200 item, it would be better load 1 item and compare if exists in the database, if yes stop otherwise load the other items

what's do you think

**Comment by marinofaggiana** [comment]
@thisIsTheFoxe You can look here for API detail (improvements) :

https://github.com/nextcloud/activity/blob/master/docs/endpoint-v2.md

**Comment by thisIsTheFoxe** [comment]
> 2. when load the view reload always the first 200 item, it would be better load 1 item and compare if exists in the database, if yes stop otherwise load the other items

Problem: I edit a file and check the details (load newest activity). Next I check all Activities (in "More" tab). Now, the first activity is loaded, skip loading new activities. But now the rest, everything in between my edit and the last activity that was saved will *never* be loaded.

**Comment by marinofaggiana** [comment]
> > 1. when load the view reload always the first 200 item, it would be better load 1 item and compare if exists in the database, if yes stop otherwise load the other items
> 
> Problem: I edit a file and check the details (load newest activity). Next I check all Activities (in "More" tab). Now, the first activity is loaded, skip loading new activities. But now the rest, everything in between my edit and the last activity that was saved will _never_ be loaded, right?

load newest activity --> will be load from newest activity TO last activity stored in our DB -> filter for METADATA , correct ?

**Comment by thisIsTheFoxe** [comment]
> > > 1. when load the view reload always the first 200 item, it would be better load 1 item and compare if exists in the database, if yes stop otherwise load the other items
> > 
> > 
> > Problem: I edit a file and check the details (load newest activity). Next I check all Activities (in "More" tab). Now, the first activity is loaded, skip loading new activities. But now the rest, everything in between my edit and the last activity that was saved will _never_ be loaded, right?
> 
> load newest activity --> will be load from newest activity TO last activity stored in our DB -> filter for METADATA , correct ?

First, not sure what you mean, I don't think it's possible it's possible to specify a range like that in the API. It'll always have to load all the newest activities.

Filtering is the problem. Because I don't think we should change the behaviour of that. If I view file activities it should only load file activities and not everything. Otherwise it might take a long time to load recent activities for that specific file. 
But that also means that there is no way of knowing whether all recent activities have been loaded, if the most recent activities match.

**Comment by marinofaggiana** [comment]
> First, not sure what you mean, I don't think it's possible it's possible to specify a range like that in the API. It'll always have to load all the newest activities.

Sure that is possible: 

GET /ocs/v2.php/apps/activity/api/v2/activity 
Parameters
since: nothing
limit: 1
return example ID: 1224

Now get your last ID from DB example 1220:

GET /ocs/v2.php/apps/activity/api/v2/activity 
Parameters
since 1224
limit 4 (1224-1220) 

works for you ?

> 
> Filtering is the problem. Because I don't think we should change the behaviour of that. If I view file activities it should only load file activities and not everything. Otherwise it might take a long time to load recent activities for that specific file. But that also means that there is no way of knowing whether all recent activities have been loaded, if the most recent activities match.

I don't understand...

**Comment by thisIsTheFoxe** [comment]
> Now get your last ID from DB example 1220:
GET /ocs/v2.php/apps/activity/api/v2/activity
Parameters
since 1224
limit 4 (1224-1220)

Yes, but that only means 1220 is the most recent. It doesn't guarantee that from 1220 all activities are actually loaded.
 
Problem: You edit some file (activityId: 1219). Next, I edit another file and check the details (load newest activity, filtered for file; activityId: 1220). Next, I check all Activities (in "More" tab). Now, the first activity in the DB has the id 1220 and the last say, 1000. So I skip reloading and never load activity 1219.

**Comment by marinofaggiana** [comment]
> > Now get your last ID from DB example 1220:
> > GET /ocs/v2.php/apps/activity/api/v2/activity
> > Parameters
> > since 1224
> > limit 4 (1224-1220)
> 
> Yes, but that only means 1220 is the most recent. It doesn't guarantee that from 1220 all activities are actually loaded.
> 
> Problem: You edit some file (activityId: 1219). Next, I edit another file and check the details (load newest activity, filtered for file; activityId: 1220). Next, I check all Activities (in "More" tab). Now, the first activity in the DB has the id 1220. So I load 1224-1220, but **never** load 1219.

because you use the filtered for file object_type in the (ACTIVITY+COMMENTS) ... so is correct but you don't have to memorize it on DB  the newest activity (it's however wrong, looking at it today,load every time 200 items).

Solution ?

- Modify DB for this two possibility (from file or not, another column maybe activityId_files)
- or don't memorized the fetch from object_type [no good ... but possible]
- or don'u use the object_type (I don't think you generate so many records / user ... #now load 200 item for nothing)

...

**Comment by marinofaggiana** [comment]
Your last push:

Schermata 2021-11-16 alle 13 59 11

where:

403 Forbidden | The offset activity belongs to a different user

**Comment by thisIsTheFoxe** [comment]
> Modify DB for this two possibility (from file or not, another column maybe activityId_files)

That would mean saving the activities twice. Wasting space.
> or don't memorized the fetch from object_type [no good ... but possible]

Yes, not good.
>or don'u use the object_type (I don't think you generate so many records / user ... #now load 200 item for nothing)


Not exactly sure what you mean. But I think it's about that:
> Because I don't think we should change the behaviour of that. If I view file activities it should only load file activities and not everything. Otherwise it might take a long time to load recent activities for that specific file.


<hr/>

> Your last push:
> 
> Schermata 2021-11-16 alle 13 59 11
> 
> where:
> 
> 403 Forbidden | The offset activity belongs to a different user

A) the screenshot shows 304 (Not Modified), not 403. And I do get those for loading avatars, not activities:

```
[Result]: failure(Alamofire.AFError.responseValidationFailed(reason: Alamofire.AFError.ResponseValidationFailureReason.unacceptableStatusCode(code: 304)))
[Request]: GET https://cloud.nextcloud.com/index.php/avatar/jos/384
    [Headers]:
        If-None-Match: "779245e54a88b76e20b621463aa03d18"
    [Body]: None
[Response]:
    [Status Code]: 304
```

**Comment by marinofaggiana** [comment]
> > Modify DB for this two possibility (from file or not, another column maybe activityId_files)
> 
> That would mean saving the activities twice. Wasting space.
> 
> > or don't memorized the fetch from object_type [no good ... but possible]
> 
> Yes, not good.
> 
> > or don'u use the object_type (I don't think you generate so many records / user ... #now load 200 item for nothing)
> 
> Not exactly sure what you mean. But I think it's about that:
> 
> > Because I don't think we should change the behaviour of that. If I view file activities it should only load file activities and not everything. Otherwise it might take a long time to load recent activities for that specific file.
> 

you propose a solution.

> > Your last push:
> > Schermata 2021-11-16 alle 13 59 11
> > where:
> > 403 Forbidden | The offset activity belongs to a different user
> 
> The screenshot shows 304 (Not Modified), not 403. And I do get those for loading avatars, not activities:
> 
> ```
> [Result]: failure(Alamofire.AFError.responseValidationFailed(reason: Alamofire.AFError.ResponseValidationFailureReason.unacceptableStatusCode(code: 304)))
> [Request]: GET https://cloud.nextcloud.com/index.php/avatar/jos/384
>     [Headers]:
>         If-None-Match: "779245e54a88b76e20b621463aa03d18"
>     [Body]: None
> [Response]:
>     [Status Code]: 304
> ```

uhh no no was from /ocs/v2.php/apps/activity/api/v2/activity/all?format=json&limit=200&previews=true&since=53334 
and 304 = end item. But something does not work ... via web I have other old items ... now I verify why.

**Comment by marinofaggiana** [comment]
> > > Modify DB for this two possibility (from file or not, another column maybe activityId_files)
> > 
> > 
> > That would mean saving the activities twice. Wasting space.
> > > or don't memorized the fetch from object_type [no good ... but possible]
> > 
> > 
> > Yes, not good.
> > > or don'u use the object_type (I don't think you generate so many records / user ... #now load 200 item for nothing)
> > 
> > 
> > Not exactly sure what you mean. But I think it's about that:
> > > Because I don't think we should change the behaviour of that. If I view file activities it should only load file activities and not everything. Otherwise it might take a long time to load recent activities for that specific file.
> 
> you propose a solution.
> 
> > > Your last push:
> > > Schermata 2021-11-16 alle 13 59 11
> > > where:
> > > 403 Forbidden | The offset activity belongs to a different user
> > 
> > 
> > The screenshot shows 304 (Not Modified), not 403. And I do get those for loading avatars, not activities:
> > ```
> > [Result]: failure(Alamofire.AFError.responseValidationFailed(reason: Alamofire.AFError.ResponseValidationFailureReason.unacceptableStatusCode(code: 304)))
> > [Request]: GET https://cloud.nextcloud.com/index.php/avatar/jos/384
> >     [Headers]:
> >         If-None-Match: "779245e54a88b76e20b621463aa03d18"
> >     [Body]: None
> > [Response]:
> >     [Status Code]: 304
> > ```
> 
> uhh no no was from /ocs/v2.php/apps/activity/api/v2/activity/all?format=json&limit=200&previews=true&since=53334 and 304 = end item. But something does not work ... via web I have other old items ... now I verify why.

Find, database dirty.

but at the end 304 we can added in footer, like the server,  the:

Schermata 2021-11-16 alle 14 16 38

**Comment by marinofaggiana** [comment]
> > you propose a solution.
> 
> Just keep a counter of the most recently loaded activity without filtering. I think that's the easiest and best compromise.

remember obligatory update (+1) the databaseSchemaVersion whenever you modify the DB

https://github.com/nextcloud/ios/blob/d2e337830d2b6c89e9c59a0516b17b54165840cf/iOSClient/NCGlobal.swift#L114

**Comment by thisIsTheFoxe** [comment]
> > > The screenshot shows 304 (Not Modified), not 403. And I do get those for loading avatars, not activities:
> > > ```
> > > [Result]: failure(Alamofire.AFError.responseValidationFailed(reason: Alamofire.AFError.ResponseValidationFailureReason.unacceptableStatusCode(code: 304)))
> > > [Request]: GET https://cloud.nextcloud.com/index.php/avatar/jos/384
> > >     [Headers]:
> > >         If-None-Match: "779245e54a88b76e20b621463aa03d18"
> > >     [Body]: None
> > > [Response]:
> > >     [Status Code]: 304
> > > ```
> > 
> > 
> > uhh no no was from /ocs/v2.php/apps/activity/api/v2/activity/all?format=json&limit=200&previews=true&since=53334 and 304 = end item. But something does not work ... via web I have other old items ... now I verify why.
> 
> Find, database dirty.
> 
> but at the end 304 we can added in footer, like the server, the:
> 
> Schermata 2021-11-16 alle 14 16 38

Good catch. But imo that's a little out of scope for this PR. Can you please make a new issue for that?

**Comment by marinofaggiana** [comment]
you can removed this unused include

https://github.com/nextcloud/ios/blob/286f97befbff496aa4bedc036ec34b876889cd56/iOSClient/Activity/NCActivity.swift#L28

File: iOSClient/Data/NCDatabase.swift [hunk 4]
```diff
@@ -107,6 +108,14 @@ class tableAccount: Object {
...
 
+class tableActivityLatestId: Object {
+    @objc dynamic var account = ""
+    @objc dynamic var mostRecentlyLoadedActivityId: Int = 0
+    override static func primaryKey() -> String {
+        return "account"
+    }
+}
+
 class tableActivity: Object, DateCompareable {
```
Comment (marinofaggiana) [inline iOSClient/Data/NCDatabase.swift:119]: small fixation, can you move this new `tableActivityLatestId` after the class `tableActivity` (sorted by letter)

**Review by marinofaggiana — CHANGES_REQUESTED** [review]

File: iOSClient/Activity/NCActivity.swift [hunk 5]
```diff
@@ -351,89 +351,132 @@ extension NCActivity: UITableViewDataSource {
...
+            objectId: metadata?.fileId,
+            objectType: objectType,
+            previews: true) { (account, activities, errorCode, errorDescription) in
+                defer { disptachGroup.leave() }
+
+                guard errorCode == 0,
+                      account == self.appDelegate.account,
+                      let activity = activities.first,
+                      activity.idActivity > recentActivityId
+                else { return }
```
Comment (marinofaggiana) [inline iOSClient/Activity/NCActivity.swift:454]: Always return at the start (DB empty) check it

Comment (marinofaggiana) [reply iOSClient/Activity/NCActivity.swift:454]: Please before update the PR, do you the test.

Comment (thisIsTheFoxe) [reply iOSClient/Activity/NCActivity.swift:454]: > Always return at the start (DB empty) check it

I'm not sure what you mean. If the DB is empty (aka `getLatestActivityId` == 0) I do return before requesting.. Feel free to elaborate..

**Review by marinofaggiana — APPROVED** [review]

**Pull request merged**